Fix v3-web-component va-segmented-progress-bar Safari focus bug

tlei123 opened the pull request:
## Summary

- Fix Safari focus bug -- focused on Continue-button instead of form-page H2 when advancing to next page.
- Team: Veteran Facing Forms

## Related issue(s)

- _Link to ticket created in va.gov-team-forms repo_
department-of-veterans-affairs/va.gov-team-forms#978

## Testing done

- Browser- & unit-tests successful

## What areas of the site does it impact?

All forms with `useCustomScrollAndFocus` prop enabled, on Safari only.

## Requested feedback

@Mottie , please be sure to review & approve.  I couldn't make you required-reviewer here, but I'm definitely gonna wait for your Approval before Merging.  Thanks much!  :)

tlei123 committed:
[WIP] Fix v3-web-component va-segmented-progress-bar Safari focus bug

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Temporarily disable dev-page-links in 21-0966 form-config

tlei123 committed:
Temporarily disable dev-page-links option in 40-0247 form-config

rhasselle-oddball commented on src/platform/forms-system/src/js/components/FormNav.jsx:
```diff
@@ -106,6 +107,11 @@ export default function FormNav(props) {
     () => {
+      // Need to provide shadowRoot for focusing on elements inside shadow-DOM
+      const root = formConfig.v3SegmentedProgressBar
+        ? document.querySelector('va-segmented-progress-bar').shadowRoot
```
the other tricky thing to consider, is sometimes the the element will be found, but the shadowRoot may not exist yet. so this type of selector is not necessarily reliable.  this is why querySelectorWithShadowRoot function was created (async)

tlei123 replied:
Thx, Robert!  Yeah, I'm making the query async in my next PR.

tlei123 replied:
[Refactors pushed](https://github.com/department-of-veterans-affairs/vets-website/pull/27784/commits/1ab1e616f18f181ba0a95b4d8a78a3d99e6046be).  Now focus code's moved out of effect-hook to an async handleFocus method where I'm using your querySelectorWithShadowRoot to first grab the shadow-host before accessing .shadowRoot.

tlei123 committed:
Fix shadow-host queries in FormNav & FormPage focus methods

tlei123 committed:
Remove remnant custom-focus references from 40-0127 form-config

tlei123 committed:
Add note to JSDoc for customScrollAndFocus method

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Update FormPage unit-tests

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Refactor appeals shared unit-test

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Restore local-dev-page-links in 21-0966 form-config

rhasselle-oddball commented on src/platform/forms-system/src/js/components/FormNav.jsx:
```diff
@@ -98,6 +100,43 @@ export default function FormNav(props) {
         ''}`;
 
+  const handleFocus = async () => {
+    let root = document.querySelector('#react-root');
```
would it be fine to just leave this as document? the fewer changes to looking up selectors the better to ensure no regressions, especially with custom HTML

rhasselle-oddball commented on src/platform/forms-system/src/js/components/FormNav.jsx:
```diff
@@ -98,6 +100,43 @@ export default function FormNav(props) {
+      root = shadowHost.shadowRoot;
+    }
+
+    return () => {
```
Focus wasn't working for me on the review page (tested mock-simple-forms-patterns). I can help test some more pages tomorrow

Mottie replied:
Hmm, this seems to be returning a function from `handleFocus`, so the useEffect call on line 137 doesn't do anything. Then it seems that [this line of code is causing focus to move to the `h2`](https://github.com/department-of-veterans-affairs/vets-website/blob/main/src/platform/forms/save-in-progress/RoutedSavableReviewPage.jsx#L32) in the footer of our page (the "Need help?" header). Maybe we need to remove this, because once this `handleFocus` function works, we don't need to manage focus in the `RoutedSavableReviewPage` component.

tlei123 replied:
Refactors made, but couldn't get the appeals forms to work on my local.  Pinged Robin to test.

rhasselle-oddball commented on src/platform/forms-system/src/js/components/FormNav.jsx:
```diff
@@ -98,6 +100,43 @@ export default function FormNav(props) {
+  // Handle focus on mount
+  useEffect(() => {
+    handleFocus();
+  }, []);
```
Is this necessary? since you already have a separate use effect for this, which will also trigger on mount + when its state changes

rhasselle-oddball commented on src/platform/forms-system/src/js/components/FormNav.jsx:
```diff
@@ -98,6 +100,43 @@ export default function FormNav(props) {
+        if (formConfig.useCustomScrollAndFocus) {
+          customScrollAndFocus(page.scrollAndFocusTarget, index);
+        } else {
+          waitForRenderThenFocus(defaultFocusSelector, root, 400);
```
maybe we still need to fallback to 'h2'. especially some custom pages. I think we need to test several kinds of custom pages to be sure.
couldn't we still use focusByOrder here? now that we're awaiting the progress bar, just wondering if we can avoid this extra 400ms time out

tlei123 replied:
Yes, just changed back to focusByOrder.

rhasselle-oddball reviewed (commented):
thanks for making these changes, left a few comments.

Mottie reviewed (commented):
Everything except for the focus on the review & submit page appears to be working in our apps

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Revert "Fix shadow-host queries in FormNav & FormPage focus methods"

This reverts commit 1ab1e616f18f181ba0a95b4d8a78a3d99e6046be.

tlei123 committed:
Revert "Revert "Fix shadow-host queries in FormNav & FormPage focus methods""

This reverts commit e60ab022373638a0d0d499149ed46a19692b92eb.

tlei123 committed:
Refactor focus methods and FormPage method-calls

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Fix FormNav focus async-code

tlei123 commented on src/platform/forms-system/test/js/containers/FormPage.unit.spec.jsx:
```diff
@@ -950,28 +950,71 @@ describe('Schemaform <FormPage>', () => {
+
       await waitFor(() => {
-        expect(document.activeElement.tagName).to.eq('H3');
+        expect(document.activeElement.tagName).to.eq('H2');
```
Deleted.

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

Mottie commented on src/platform/forms-system/src/js/components/FormNav.jsx:
```diff
@@ -111,35 +107,14 @@ export default function FormNav(props) {
-          // h2 fallback for confirmation page
-          focusByOrder([defaultFocusSelector, 'h2']);
+      handleFocus(page, formConfig, index).then(() => {
+        const progressBar = document.querySelector('va-segmented-progress-bar');
+        if (progressBar) {
+          scrollTo(document.querySelector('va-segmented-progress-bar'));
         }
-      };
+      });
```
Having this `.then()` here seems like no matter what `handleFocus` does with scrolling (it calls a `customScrollAndFocus` if it exists), the scroll will get overridden if there is a `va-segmented-progress-bar` on the page. Maybe scroll first, then call `handleFocus`?

Mottie commented on src/platform/forms-system/src/js/containers/FormPage.jsx:
```diff
@@ -24,21 +24,40 @@ import {
+import { querySelectorWithShadowRoot } from '../../../../utilities/ui/webComponents';
 
-function focusForm(route, index) {
+async function focusForm(route, index) {
```
This `focusForm` appears to have duplicate code from the `getFocusTargetRoot` and `handleFocus` functions? Can these be combined?

tlei123 committed:
Commit 20-10207 file-deletions left over from main-branch merge

tlei123 committed:
Update 21-0966 local-dev-links-config and local-mock-api responses

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Update 21-0247 local-dev-links config and local-mock-api responses

tlei123 committed:
Commit 20-10207 file-deletions left-over from main-merge

tlei123 committed:
WIP Refactor FormNav scroll and focus

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Refactor FormPage focusForm method

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Update 40-0247 SIP mock-api-responses

Mottie reviewed (dismissed):
LGTM

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

Mottie commented:
Ok, I had to un-approve. I forgot I was on your fork and I was testing our form and I found that the focus jumped back to the H3 (`scrollAndFocusTarget`) every time I typed a character.

![focus](https://github.com/department-of-veterans-affairs/vets-website/assets/136959/db6bb61b-3d55-42b3-b13f-80d505365724)

tlei123 committed:
Fix unwanted field-input-triggered focusing in FormNav effect-hook

tlei123 commented:
> Ok, I had to un-approve. I forgot I was on your fork and I was testing our form and I found that the focus jumped back to the H3 (`scrollAndFocusTarget`) every time I typed a character.
> 
> ![focus](https://private-user-images.githubusercontent.com/136959/306440467-db6bb61b-3d55-42b3-b13f-80d505365724.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0ODA2MDcsIm5iZiI6MTcwODQ4MDMwNywicGF0aCI6Ii8xMzY5NTkvMzA2NDQwNDY3LWRiNmJiNjFiLTNkNTUtNDJiMy1iMTNmLTgwZDUwNTM2NTcyNC5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMVQwMTUxNDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNjhlNGNkNjI2OWIwMjQ3NTE5MjUxZTQ5MmM4M2VmYzZiZGJmMDJkYTJhMDE4NGM0OWI1YWFmNWE3MzgzZmQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9omT0VzuFIPTAF4Nwl4_HzAFKKFAvshJQeT7fDJHhp8) [ ![focus](https://private-user-images.githubusercontent.com/136959/306440467-db6bb61b-3d55-42b3-b13f-80d505365724.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0ODA2MDcsIm5iZiI6MTcwODQ4MDMwNywicGF0aCI6Ii8xMzY5NTkvMzA2NDQwNDY3LWRiNmJiNjFiLTNkNTUtNDJiMy1iMTNmLTgwZDUwNTM2NTcyNC5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMVQwMTUxNDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNjhlNGNkNjI2OWIwMjQ3NTE5MjUxZTQ5MmM4M2VmYzZiZGJmMDJkYTJhMDE4NGM0OWI1YWFmNWE3MzgzZmQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9omT0VzuFIPTAF4Nwl4_HzAFKKFAvshJQeT7fDJHhp8) ](https://private-user-images.githubusercontent.com/136959/306440467-db6bb61b-3d55-42b3-b13f-80d505365724.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0ODA2MDcsIm5iZiI6MTcwODQ4MDMwNywicGF0aCI6Ii8xMzY5NTkvMzA2NDQwNDY3LWRiNmJiNjFiLTNkNTUtNDJiMy1iMTNmLTgwZDUwNTM2NTcyNC5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMVQwMTUxNDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNjhlNGNkNjI2OWIwMjQ3NTE5MjUxZTQ5MmM4M2VmYzZiZGJmMDJkYTJhMDE4NGM0OWI1YWFmNWE3MzgzZmQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9omT0VzuFIPTAF4Nwl4_HzAFKKFAvshJQeT7fDJHhp8) [ ](https://private-user-images.githubusercontent.com/136959/306440467-db6bb61b-3d55-42b3-b13f-80d505365724.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0ODA2MDcsIm5iZiI6MTcwODQ4MDMwNywicGF0aCI6Ii8xMzY5NTkvMzA2NDQwNDY3LWRiNmJiNjFiLTNkNTUtNDJiMy1iMTNmLTgwZDUwNTM2NTcyNC5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMVQwMTUxNDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNjhlNGNkNjI2OWIwMjQ3NTE5MjUxZTQ5MmM4M2VmYzZiZGJmMDJkYTJhMDE4NGM0OWI1YWFmNWE3MzgzZmQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9omT0VzuFIPTAF4Nwl4_HzAFKKFAvshJQeT7fDJHhp8)

@Mottie , this is now fixed.  Please retest.

"No good deed goes unpunished!"  Followed linter suggestion to add more dependencies to FormNav's effect-hook, which turned out to trigger this input-triggered focus.  Removed those added ones, and input-triggered focusing's no longer occuring here at my end.

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

Mottie commented on src/platform/forms-system/src/js/components/FormNav.jsx:
```diff
@@ -111,35 +107,15 @@ export default function FormNav(props) {
-          focusByOrder([defaultFocusSelector, 'h2']);
-        }
-      };
+      scrollTo('vaSegmentedProgressBar', { offset: -20 });
```
Should this `scrollTo` be moved into the `handleFormNavFocus` function?

Mottie reviewed (dismissed):
Thanks for the update! It's working great now!

jeana-adhoc commented:
@tlei123 - This works great for me! @rhasselle-oddball - Do you have NVDA or JAWS on your windows machine, it'd be great to test this with one of those just to make sure we didn't cause any regressions.

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Remove console.logs

tlei123 committed:
Add new unit-tests for getFormNavFocusTargetRoot & handleFormNavFocus

Mottie reviewed (dismissed):
I saw a typo `dispath` vs `dispatch`; but it's minor

rhasselle-oddball commented on src/applications/simple-forms/20-10207/pages/evidencePOWConfinement.js:
it looks like these ones are being correctly deleted

rhasselle-oddball commented on src/applications/simple-forms/20-10207/pages/evidencePowConfinement.js:
but I think these ones are accidentally being deleted? it looks like the form config still depends on this, maybe this is why the unit test are failing.

rhasselle-oddball reviewed (commented):
tested financial-status-report/appeals/pensions, and seems like its working fine

tlei123 committed:
Fix customScrollAndFocus for va-segmented-progress-bar

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Fix anomalous PP10207 file-deletions

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

rhasselle-oddball reviewed (dismissed):
testing checks out for me 👍left a few nit commits if there is another patch.  I think the next hurdle will just be getting all the unit tests to pass.

tlei123 committed:
Update 21-10210 utils unit-tests

tlei123 committed:
Refactor FormNav & FormPage scroll/focus code.

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Revert "Refactor FormNav & FormPage scroll/focus code."

This reverts commit b92eadf2d867d2bc8d04b87781be09b39e6c769a.

tlei123 committed:
Revert "Revert "Refactor FormNav & FormPage scroll/focus code.""

This reverts commit 128d0fb769e87b789b1ff67d9415fd33c524e1df.

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
CI-skip secure-messaging-compose e2e-spec due to flakiness

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

ToddWebDev commented:
When I check out this branch locally and test it with the Pension benefits application (/pension/application/527EZ), the issue still persists on Safari. Chrome also appears to have the issue too, where it didn't before. I've made sure to check out the latest commit b059f87d22bba5a256b8b8ad6aa2c78670a8be03. Are there any steps I'm missing to run the fix locally?

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

ToddWebDev commented:
If we switch the Pension benefits application uiSchema titles from `'ui:title':  '...'` to the title web component pattern: `...titleUI('...')`, then it seems to be included with this fix. I am testing now.

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

tlei123 committed:
Merge branch 'main' into simple-forms/978-safari-focus-2

oleksii-morgun commented on src/applications/mhv/secure-messaging/tests/e2e/secure-messaging-compose.cypress.spec.js:
```diff
@@ -4,7 +4,10 @@ import PatientComposePage from './pages/PatientComposePage';
+// Skip in CI due to flakiness
+const testSuite = Cypress.env('CI') ? describe.skip : describe;
+
+testSuite('Secure Messaging Compose', () => {
```
@fazilqa for your awareness. We may have to review what is causing it to be flaky

tlei123 commented:
Closing this PR.  Left un-merged since June 2024.
